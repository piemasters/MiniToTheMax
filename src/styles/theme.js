export default {
  colors: {
    text: '#000',
    muted: '#999',
  },
};
