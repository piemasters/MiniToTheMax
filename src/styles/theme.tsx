export interface Theme {
  colors: {
    text: string;
    muted: string;
  };
}

const appTheme: Theme = {
  colors: {
    text: '#000',
    muted: '#999',
  },
};

export default appTheme;
