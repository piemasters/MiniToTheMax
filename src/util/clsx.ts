/* eslint-disable @typescript-eslint/no-unused-expressions */
/* eslint-disable @typescript-eslint/no-explicit-any */
export type ClassValue =
  | ClassArray
  | ClassDictionary
  | string
  | number
  | null
  | boolean
  | undefined;
type ClassDictionary = Record<string, any>;
type ClassArray = ClassValue[];

function toVal(mix: string | number | boolean | unknown[] | ClassDictionary) {
  let k,
    y,
    str = '';

  if (typeof mix === 'string' || typeof mix === 'number') {
    str += mix;
  } else if (typeof mix === 'object') {
    if (Array.isArray(mix)) {
      const len = mix.length;
      for (k = 0; k < len; k++) {
        if (mix[k]) {
          if ((y = toVal(mix[k]))) {
            str && (str += ' ');
            str += y;
          }
        }
      }
    } else {
      for (y in mix) {
        if (mix[y]) {
          str && (str += ' ');
          str += y;
        }
      }
    }
  }

  return str;
}

export function clsx(...args: ClassValue[]): string {
  let i = 0,
    tmp,
    x,
    str = '';
  const len = args.length;
  for (; i < len; i++) {
    if ((tmp = args[i])) {
      if ((x = toVal(tmp))) {
        str && (str += ' ');
        str += x;
      }
    }
  }
  return str;
}

export default clsx;
