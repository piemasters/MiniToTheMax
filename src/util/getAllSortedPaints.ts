import { PaintDetails } from '../types';
import { textToId } from './textToId';
import { sortPaints } from './sortPaints';
import {
  airPaints,
  basePaints,
  contrastPaints,
  dryPaints,
  layerPaints,
  shadePaints,
  sprayPaints,
  technicalPaints,
} from '../data/paints/citadel';
import { gameColorPaints } from '../data/paints';

export const getAllSortedPaints = (): PaintDetails[] => {
  const allPaints: PaintDetails[] = [
    // citadel
    ...airPaints,
    ...basePaints,
    ...contrastPaints,
    ...dryPaints,
    ...layerPaints,
    ...shadePaints,
    ...sprayPaints,
    ...technicalPaints,
    // vallejo
    ...gameColorPaints,
  ];

  // Remove duplicates using a Set
  const uniquePaints = Array.from(
    new Set(
      allPaints.map(
        (paint) =>
          `${textToId(paint.company)}_${textToId(paint.name)}_${textToId(paint.category)}`
      )
    )
  ).map((uniqueKey) =>
    allPaints.find(
      (paint) =>
        `${textToId(paint.company)}_${textToId(paint.name)}_${textToId(paint.category)}` ===
        uniqueKey
    )
  ) as PaintDetails[];

  // Sort the paints using the sortPaints function
  return uniquePaints.sort(sortPaints);
};
