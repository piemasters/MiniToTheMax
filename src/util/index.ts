export * from './pagination';
export * from './paintFilters';
