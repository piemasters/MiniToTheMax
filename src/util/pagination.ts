interface getPagesProps {
  currentPage: number;
  numPages: number;
  maxPages: number;
}

export const getPages = ({
  currentPage,
  numPages,
  maxPages,
}: getPagesProps) => {
  let startPage: number;
  let endPage: number;
  if (numPages <= maxPages) {
    // Show all pages
    startPage = 1;
    endPage = numPages;
  } else {
    if (currentPage <= Math.floor(maxPages / 2)) {
      // Left end [ 1 [2] 3 4 ] || [ 1 [2] 3 ]
      startPage = 1;
      endPage = maxPages;
    } else if (numPages - currentPage < Math.floor(maxPages / 2) + 1) {
      // Right end [ 7 [8] 9 10 ] || [ 8 [9] 10 ]
      startPage = numPages - maxPages + 1;
      endPage = numPages;
    } else {
      // Middle [ 4 5 [6] 7 8 ] || [ 4 [5] 6 7 ]
      const even = maxPages % 2 === 0;
      startPage = even
        ? currentPage - Math.floor(maxPages / 2) + 1
        : currentPage - Math.floor(maxPages / 2);
      endPage = even
        ? currentPage + Math.ceil(maxPages / 2)
        : currentPage + Math.ceil(maxPages / 2) - 1;
    }
  }

  // Return an array of pages to repeat
  return [...Array(endPage + 1 - startPage).keys()].map((i) => startPage + i);
};
