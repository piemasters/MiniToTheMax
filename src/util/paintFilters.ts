import {
  AllPaintCategories,
  PaintAvailabilities,
  PaintColors,
  PaintCompanies,
  PaintFilters,
  PaintTypes,
} from '../types';

// Build filters from types, with all values set to true by default

export const companyFilters: PaintFilters = PaintCompanies.reduce(
  (filters, company) => {
    filters[company] = true;
    return filters;
  },
  {} as PaintFilters
);

export const categoryFilters: PaintFilters = [...AllPaintCategories]
  .sort((a, b) => a.localeCompare(b))
  .reduce((filters, category) => {
    filters[category] = true;
    return filters;
  }, {} as PaintFilters);

export const colorFilters: PaintFilters = PaintColors.reduce(
  (filters, color) => {
    filters[color] = true;
    return filters;
  },
  {} as PaintFilters
);

export const typeFilters: PaintFilters = PaintTypes.reduce((filters, type) => {
  filters[type] = true;
  return filters;
}, {} as PaintFilters);

export const availabilityFilters: PaintFilters = PaintAvailabilities.reduce(
  (filters, availability) => {
    filters[availability] = true;
    return filters;
  },
  {} as PaintFilters
);
