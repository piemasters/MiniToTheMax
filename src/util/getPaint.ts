import {
  CitadelPaintCategories,
  PaintCategory,
  PaintDetails,
  VallejoPaintCategories,
} from '../types';
import { getAllSortedPaints } from './getAllSortedPaints';

const allPaints = getAllSortedPaints();

const paintCategoryPriority = [
  ...CitadelPaintCategories,
  ...VallejoPaintCategories,
] as const;

type category = (typeof paintCategoryPriority)[number];

export const getPaint = (
  name: string,
  category?: PaintCategory
): { paint?: PaintDetails; error?: string; type?: string } => {
  const paintMatches = allPaints.filter((paint) => paint.name === name);

  if (paintMatches.length === 0) {
    return { error: 'Incorrect Paint Name', type: name };
  }

  if (paintMatches.length === 1) {
    return { paint: paintMatches[0] };
  }

  if (category) {
    const categoryMatches = paintMatches.filter(
      (paint) => paint.category === category
    );
    if (categoryMatches.length === 0) {
      return { error: 'Incorrect Paint Category', type: category };
    }
    return { paint: categoryMatches[0] };
  }

  const getCategoryPriority = (category: PaintCategory) => {
    const index = paintCategoryPriority.indexOf(category);
    return index === -1 ? Infinity : index;
  };
  paintMatches.sort(
    (a, b) => getCategoryPriority(a.category) - getCategoryPriority(b.category)
  );

  return { paint: paintMatches[0] };
};
