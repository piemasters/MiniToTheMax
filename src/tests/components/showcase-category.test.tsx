import React from 'react';
import ShowcaseCategory from '../../components/showcase-category';
import { getByTestId } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';
import theme from '../../styles/theme';
import { renderWithTransitionProvider } from '../util/transition-provider';

describe('ShowcaseCategory', () => {
  test('renders correctly', () => {
    const { container } = renderWithTransitionProvider(
      <ThemeProvider theme={theme}>
        <ShowcaseCategory
          img={{
            aspectRatio: 2,
            src: 'cover.jpeg',
            srcSet: '',
            sizes: '(max-width: 800px) 100vw, 800px',
          }}
          slug={'/showcase/board-games'}
          title={'Board Games'}
        />
      </ThemeProvider>
    );

    expect(getByTestId(container, 'showcase-category')).toHaveTextContent(
      'Board Games'
    );
    expect(container).toMatchSnapshot();
  });
});
