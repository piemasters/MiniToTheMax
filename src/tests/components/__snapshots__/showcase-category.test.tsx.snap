// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowcaseCategory renders correctly 1`] = `
<div>
  <div
    css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
    data-testid="showcase-category"
  >
    <a
      bg="#eb1d23"
      css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
      data-testid="page-link-cover"
      direction="up"
      duration="1"
      href="/showcase/board-games"
    >
      <div
        css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
        data-testid="cover-image"
      >
        <div
          class=" gatsby-image-wrapper"
          style="position: relative; overflow: hidden;"
        >
          <div
            aria-hidden="true"
            style="width: 100%; padding-bottom: 50%;"
          />
          <picture>
            <source
              sizes="(max-width: 800px) 100vw, 800px"
              srcset=""
            />
            <img
              alt=""
              loading="lazy"
              sizes="(max-width: 800px) 100vw, 800px"
              src="cover.jpeg"
              srcset=""
              style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; object-fit: cover; object-position: center; opacity: 0; transition: opacity 500ms;"
            />
          </picture>
          <noscript>
            &lt;picture&gt;&lt;source srcset="" sizes="(max-width: 800px) 100vw, 800px" /&gt;&lt;img loading="lazy" sizes="(max-width: 800px) 100vw, 800px" src="cover.jpeg" alt="" style="position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
          </noscript>
        </div>
        <h2
          css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
        >
          Board Games
        </h2>
      </div>
    </a>
  </div>
</div>
`;
